components/layout_2020/flow/float.rs

```diff
@@ -2,29 +2,484 @@
+                    None => new_band.right = Some(new_value),
+                    Some(ref mut old_value) => *old_value = old_value.min(new_value),
+                },
+            }
+        }
+
+        let new_left = match self.left.0 {
+            None => FloatBandLink(None),
+            Some(ref old_left) if range.start < new_band.top => {
+                FloatBandLink(Some(old_left.set_range(range, side, new_value)))
+            },
+            Some(ref old_left) => FloatBandLink(Some((*old_left).clone())),
+        };
+
+        let new_right = match self.right.0 {
+            None => FloatBandLink(None),
+            Some(ref old_right) if range.end > new_band.top => {
+                FloatBandLink(Some(old_right.set_range(range, side, new_value)))
+            },
+            Some(ref old_right) => FloatBandLink(Some((*old_right).clone())),
+        };
+
+        Arc::new(FloatBandNode {
+            band: new_band,
+            left: new_left,
+            right: new_right,
+            level: self.level,
+        })
+    }
+}
+
+impl FloatBandLink {
+    /// Returns the first band whose top is less than or equal to the given `block_position`.
+    fn find(&self, block_position: Length) -> Option<FloatBand> {
+        let this = match self.0 {
+            None => return None,
+            Some(ref node) => node,
+        };
+
+        if block_position < this.band.top {
+            return this.left.find(block_position);
+        }
+
+        // It's somewhere in this subtree, but we aren't sure whether it's here or in the right
+        // subtree.
+        if let Some(band) = this.right.find(block_position) {
+            return Some(band);
+        }
+
+        Some(this.band.clone())
+    }
+
+    /// Returns the first band whose top is strictly greater than the given `block_position`.
+    fn find_next(&self, block_position: Length) -> Option<FloatBand> {
+        let this = match self.0 {
+            None => return None,
+            Some(ref node) => node,
+        };
+
+        if block_position >= this.band.top {
+            return this.right.find_next(block_position);
+        }
+
+        // It's somewhere in this subtree, but we aren't sure whether it's here or in the left
+        // subtree.
+        if let Some(band) = this.left.find_next(block_position) {
+            return Some(band);
+        }
+
+        Some(this.band.clone())
+    }
+
+    // Inserts a new band into the tree. If the band has the same level as a pre-existing one,
+    // replaces the existing band with the new one.
+    fn insert(&self, band: FloatBand) -> FloatBandLink {
+        let mut this = match self.0 {
+            None => return FloatBandLink(Some(Arc::new(FloatBandNode::new(band)))),
+            Some(ref this) => (**this).clone(),
+        };
+
+        if band.top < this.band.top {
+            this.left = this.left.insert(band);
+            return FloatBandLink(Some(Arc::new(this))).skew().split();
```

Comment:
aren't skew and split inverses of each other?

Reply:
No, because the level checks are different. Skew operates if level(T) = level(L) and split operates if level(T) = level(X).

Reply:
Oh, I didn't notice the level checks